### PR TITLE
community: Added propagation of document metadata from O365BaseLoader

### PR DESCRIPTION
**Pull request by MacanPN**
**Description:**
- Added propagation of document metadata from O365BaseLoader to FileSystemBlobLoader (O365BaseLoader uses FileSystemBlobLoader under the hood). 
  - This is done by passing dictionary `metadata_dict`: key=filename and value=dictionary containing document's metadata
- Modified `FileSystemBlobLoader` to accept the `metadata_dict`, use `mimetype` from it (if available) and pass metadata further into blob loader.

**Issue:**
- `O365BaseLoader` under the hood downloads documents to temp folder and then uses `FileSystemBlobLoader` on it.
- However metadata about the document in question is lost in this process. In particular:
  - `mime_type`: `FileSystemBlobLoader` guesses `mime_type` from the file extension, but that does not work 100% of the time. 
  - `web_url`: this is useful to keep around since in RAG LLM we might want to provide link to the source document. In order to work well with document parsers, we pass the `web_url` as `source` (`web_url` is ignored by parsers, `source` is preserved)

**Dependencies:**
None

**Twitter handle:**
@martintriska1

Please review @baskaryan

### CODE REVIEW

**Comment by MacanPN** [comment]
@eyurtsev @baskaryan The PR is ready for review and merging! :)

**Comment by MacanPN** [comment]
@eyurtsev @baskaryan Can you please provide me with a feedback and way to move forward with this PR? Thanks!

**Comment by petergoldstein** [comment]
This PR is required for an internal deployment in a large enterprise environment.  We'd really like to see it merged.

Given that all checks pass, and the changes are relatively minor, what would help get it merged into an official version?  Any thoughts @hwchase17 ?

Happy to discuss the use case in more detail if that helps.  Thanks.

**Comment by MacanPN** [comment]
!ping @eyurtsev @baskaryan @hwchase17

File: libs/community/langchain_community/document_loaders/blob_loaders/file_system.py [hunk 1]
```diff
@@ -58,6 +68,7 @@ def __init__(
         glob: str = "**/[!.]*",
         exclude: Sequence[str] = (),
         suffixes: Optional[Sequence[str]] = None,
+        metadata_dict: Optional[Dict[str, Dict[str, Any]]] = None,
```
Comment (eyurtsev) [inline libs/community/langchain_community/document_loaders/blob_loaders/file_system.py:71]: This logic should be handled outside the file system blob loader. The code that is using the BlobLoader can inspect the the blobs and add metadata to them based on the path associated with them.

**Review by eyurtsev — CHANGES_REQUESTED** [review]

**Comment by MacanPN** [comment]
@eyurtsev I've modified the PR to only work within base_o365 and sharepoint classes. Please take a look

File: libs/community/langchain_community/document_loaders/base_o365.py [hunk 2]
```diff
@@ -108,14 +109,27 @@ def _load_from_folder(self, folder: Folder) -> Iterable[Blob]:
...
+                        metadata_dict[file.name]["mime_type"] = file.mime_type
+                        metadata_dict[file.name]["created"] = file.created
+                        metadata_dict[file.name]["modified"] = file.modified
+                        metadata_dict[file.name]["created_by"] = str(file.created_by)
+                        metadata_dict[file.name]["modified_by"] = str(file.modified_by)
+                        metadata_dict[file.name]["description"] = file.description
+
             loader = FileSystemBlobLoader(path=temp_dir)
-            yield from loader.yield_blobs()
+            for document in loader.yield_blobs():
```
Comment (eyurtsev) [inline libs/community/langchain_community/document_loaders/base_o365.py:129]: document != blob

```suggestion
            for blob in loader.yield_blobs():
```

File: libs/community/langchain_community/document_loaders/base_o365.py [hunk 3]
```diff
@@ -108,14 +109,27 @@ def _load_from_folder(self, folder: Folder) -> Iterable[Blob]:
...
+                if document.path and hasattr(document.path, "name"):
```
Comment (eyurtsev) [inline libs/community/langchain_community/document_loaders/base_o365.py:130]: Why is there a hasattr on document.path?  Blob.path is always defined.

File: libs/community/langchain_community/document_loaders/base_o365.py [hunk 4]
```diff
@@ -108,14 +109,27 @@ def _load_from_folder(self, folder: Folder) -> Iterable[Blob]:
...
         items = folder.get_items()
+        metadata_dict: Dict[str, Dict[str, Any]] = {}
         with tempfile.TemporaryDirectory() as temp_dir:
             os.makedirs(os.path.dirname(temp_dir), exist_ok=True)
             for file in items:
                 if file.is_file:
                     if file.mime_type in list(file_mime_types.values()):
                         file.download(to_path=temp_dir, chunk_size=self.chunk_size)
+                        metadata_dict[file.name] = {}
+                        metadata_dict[file.name]["source"] = file.web_url
```
Comment (eyurtsev) [inline libs/community/langchain_community/document_loaders/base_o365.py:120]: A better way to write this to reduce code duplication:

```
metadata_dict[file.name] = {
    "source" : file.web_url,
    "mime_type" : file.mime_type,
    "created" : file.created,
    "modified" : file.modified,
    "created_by" : str(file.created_by),
    "modified_by" : str(file.modified_by),
    "description" : file.description,
}
```

**Comment by eyurtsev** [comment]
@MacanPN going to push a few changes to your PR in a bit

**Review by eyurtsev — APPROVED** [review]

**Comment by eyurtsev** [comment]
@MacanPN sorry for the delay, pushed some aesthetic changes through, let me know if looks good, if so i'll merge.

**Comment by MacanPN** [comment]
@eyurtsev Changes look good. Thanks! I've now also resolved a conflict with `master`. Feel free to merge!
_Full speed ahead Mr. Spock :)_

**Comment by radvanyimome** [comment]
Is this change included in the 0.2.2 releases?
I'm still getting

"ValueError: data=None mimetype=None encoding='utf-8' path=PosixPath('/tmp/tmprxucxkc7/filename.docx') metadata={} does not have a mimetype"

**Pull request merged**